Adding a scraper for the US Coast Guard Court of Criminal Appeals.

EvandoBlanco opened the pull request:
This is my first time doing an open source contribution. Please let me know if there's anything I need to edit/update.

arderyp commented on juriscraper/opinions/united_states/federal_special/uscg_ccap.py:
```diff
@@ -0,0 +1,46 @@
+        # encoding them to be a good citizen.
+        url_xpath = "//tr//td[1]/a/@href"
+
+        return [requote_uri(url) for url in self.html.xpath(url_xpath)]
```
cool `requote_uri`!  I wasn't aware of this.

EvandoBlanco replied:
Neither was I haha.

arderyp replied:
were the URLs parsed from the page wonky without running them through this function?  If so, @mlissner, I'm wondering if it might be helpful to run this function against all urls up in `AbstractSite`

EvandoBlanco replied:
The URLs contain spaces. Isn't necessarily an issue and the links resolve in Firefox with and without encoding. It's just a nice to have in case someone needs to work with them programmatically or if some browser doesn't handle it.

arderyp commented on juriscraper/opinions/united_states/federal_special/uscg_ccap.py:
```diff
@@ -0,0 +1,46 @@
+            "CGCCA-Opinions/smdpage15701/{}"
+        )
+
+    def __format_url(self):
```
am I missing something?  Does this get called somewhere? Is this related to python's magic `__format__` method?

EvandoBlanco replied:
That is 100% something I forgot to clean up with a typo in the method name. I will clean that up.

EvandoBlanco replied:
I cleaned it up. Leaving ` _format_url()` in (with a correct name) to make implementing `_download_backwards()` easy if it's needed.

arderyp replied:
Ok, now I see the method being used, so it makes sense.  

Just FYI, `_download_backwards` will not run on this scraper.  The default behavior is nothing happens when back-scraping a site (see [here](https://github.com/freelawproject/juriscraper/blob/master/juriscraper/AbstractSite.py#L415-L417)).  

You can test by running the following from within the repo root on your terminal:

```
python sample_caller.py -c juriscraper.opinions.united_states.federal_special.uscg_ccap --backscrape
```

So, if you want a back scraper to be available, you will need to override and implement `_download_backwards`, and will also need to implement the `self.back_scrape_iterable` property in the `__init__`.  The GA scraper is a decent example: https://github.com/freelawproject/juriscraper/blob/master/juriscraper/opinions/united_states/state/ga.py

EvandoBlanco replied:
Thanks, I'll probably do that later this week, just wanted to dip a toe in haha.

arderyp reviewed (approved):
looks good to me, but will wait on @mlissner or @flooie to review.  Congrats on your first open source contribution :metal:

arderyp commented:
@flooie go ahead and merge if it looks good to you

flooie commented:
Hey @EvandoBlanco This is a great first effort, but before I do a review of the code I wanted to take note of a few things that need to be looked into.  

The minor thing is and @arderyp @mlissner please feel free to weigh in but I think the name should be `uscg_cca` not `uscg_ccap`.  I did a quick check and I don't think we add the "p" in any appellate scraper.  

The larger thing is that the scraper doesn't appear to be working perfectly quite yet.  if you look at the example JSON - the first item is keyed to this row on the website.

<img width="716" alt="Screen Shot 2021-03-30 at 2 15 13 PM" src="https://user-images.githubusercontent.com/6464529/113036482-5f805300-9162-11eb-9eb9-afdc946c6547.png">

`{
    "case_dates": "2020-09-21",
    "case_names": "Docket No. 1469",
    "download_urls": "https://media.defense.gov/2020/Sep/21/2002501702/-1/-1/0/UNITED%20STATES%20V.%20WEISER%20__M.J.%20___.PDF",
    "precedential_statuses": "Published",
    "blocked_statuses": false,
    "date_filed_is_approximate": false,
    "case_name_shorts": "Docket No. 1469"
  }`

1) It looks like case name, and case name shorts both have docket information in them and we aren't getting case title anywhere. 

2) The scraper is assigning published to every opinion scraped but you can see from the website that they denote Published and Unpublished in the title field.  It looks like they are either published in the Military Journal or Unpublished opinions.  

<img width="772" alt="Screen Shot 2021-03-30 at 2 25 24 PM" src="https://user-images.githubusercontent.com/6464529/113037841-ceaa7700-9163-11eb-89f1-6ece686f5520.png">

3) Speaking of publishing - it looks like they are providing us the exact citation (published or unpublished) with the title, we should grab that as well.

I believe we are have both a 

`_get_neutral_citations` and a `_get_west_citations` methods in OpinionSite.

I'm not sure if we are using the get_west_citations any more but I think it would be wise to capture all the citations.

flooie commented on juriscraper/opinions/united_states/federal_special/__init__.py:
```diff
@@ -13,4 +13,5 @@
     "uscfc_u",
     "uscfc_vaccine",
     "uscfc_vaccine_u",
+    "uscg_ccap",
```
@mlissner I think we should rename this to `uscg_cca`

mlissner replied:
All of your notes look good to me, @flooie. I don't know if we've documented this, but if we put underscores in the file name, everything after the first underscore is ignored by CourtListener. So if we used `uscg_cca`, the court on CL would be `uscg`, which I don't think is what we want. I think on CL we'd want the court ID to be `uscgcca`, so the filename should be `uscgcca.py`.

arderyp commented:
> Hey @EvandoBlanco This is a great first effort, but before I do a review of the code I wanted to take note of a few things that need to be looked into.
> 
> The minor thing is and @arderyp @mlissner please feel free to weigh in but I think the name should be `uscg_cca` not `uscg_ccap`. I did a quick check and I don't think we add the "p" in any appellate scraper.
> 
> The larger thing is that the scraper doesn't appear to be working perfectly quite yet. if you look at the example JSON - the first item is keyed to this row on the website.
> 
> <img alt="Screen Shot 2021-03-30 at 2 15 13 PM" width="716" src="https://user-images.githubusercontent.com/6464529/113036482-5f805300-9162-11eb-9eb9-afdc946c6547.png">
> 
> `{ "case_dates": "2020-09-21", "case_names": "Docket No. 1469", "download_urls": "https://media.defense.gov/2020/Sep/21/2002501702/-1/-1/0/UNITED%20STATES%20V.%20WEISER%20__M.J.%20___.PDF", "precedential_statuses": "Published", "blocked_statuses": false, "date_filed_is_approximate": false, "case_name_shorts": "Docket No. 1469" }`
> 
>     1. It looks like case name, and case name shorts both have docket information in them and we aren't getting case title anywhere.
> 
>     2. The scraper is assigning published to every opinion scraped but you can see from the website that they denote Published and Unpublished in the title field.  It looks like they are either published in the Military Journal or Unpublished opinions.
> 
> 
> <img alt="Screen Shot 2021-03-30 at 2 25 24 PM" width="772" src="https://user-images.githubusercontent.com/6464529/113037841-ceaa7700-9163-11eb-89f1-6ece686f5520.png">
> 
>     1. Speaking of publishing - it looks like they are providing us the exact citation (published or unpublished) with the title, we should grab that as well.
> 
> 
> I believe we are have both a
> 
> `_get_neutral_citations` and a `_get_west_citations` methods in OpinionSite.
> 
> I'm not sure if we are using the get_west_citations any more but I think it would be wise to capture all the citations.

fantastic... you obviously tested where I didn't :facepalm:

flooie commented:
> @flooie Will do, I'll take some time and update the scraper.

This is still great.  I've been hoping to eventually get to the military Discharge Review Boards and Tribal courts eventually, so it's great to add new military courts!

EvandoBlanco commented:
> 
> 
> Hey @EvandoBlanco This is a great first effort, but before I do a review of the code I wanted to take note of a few things that need to be looked into.
> 
> The minor thing is and @arderyp @mlissner please feel free to weigh in but I think the name should be `uscg_cca` not `uscg_ccap`. I did a quick check and I don't think we add the "p" in any appellate scraper.
> 
> The larger thing is that the scraper doesn't appear to be working perfectly quite yet. if you look at the example JSON - the first item is keyed to this row on the website.
> 
> <img alt="Screen Shot 2021-03-30 at 2 15 13 PM" width="716" src="https://user-images.githubusercontent.com/6464529/113036482-5f805300-9162-11eb-9eb9-afdc946c6547.png">
> 
> `{ "case_dates": "2020-09-21", "case_names": "Docket No. 1469", "download_urls": "https://media.defense.gov/2020/Sep/21/2002501702/-1/-1/0/UNITED%20STATES%20V.%20WEISER%20__M.J.%20___.PDF", "precedential_statuses": "Published", "blocked_statuses": false, "date_filed_is_approximate": false, "case_name_shorts": "Docket No. 1469" }`
> 
>     1. It looks like case name, and case name shorts both have docket information in them and we aren't getting case title anywhere.
> 
>     2. The scraper is assigning published to every opinion scraped but you can see from the website that they denote Published and Unpublished in the title field.  It looks like they are either published in the Military Journal or Unpublished opinions.
> 
> 
> <img alt="Screen Shot 2021-03-30 at 2 25 24 PM" width="772" src="https://user-images.githubusercontent.com/6464529/113037841-ceaa7700-9163-11eb-89f1-6ece686f5520.png">
> 
>     1. Speaking of publishing - it looks like they are providing us the exact citation (published or unpublished) with the title, we should grab that as well.
> 
> 
> I believe we are have both a
> 
> `_get_neutral_citations` and a `_get_west_citations` methods in OpinionSite.
> 
> I'm not sure if we are using the get_west_citations any more but I think it would be wise to capture all the citations.

FYI, an InsanityException gets raised if the # of citations does not match the number of items. Is there a convention for this? Should I return a None if there's no info in the name? An empty string?

flooie commented:
Yes - I think thats right.  The scraper requires that the lengths of all lists come back even.  So if you are grabbing a field - I think none makes sense its not there.

mlissner commented:
Seconding what @flooie said. The insanity exception is to tell you that your data isn't sane (get it?). The idea is that you have various lists of data that will get merged with the first item from each coming together to form a case, then the second, and so forth. And if the lists are differing lengths, that's a big problem because they won't merge together the right way. I think @arderyp actually came up with a better object design after we worked with this terrible concept for many years. Maybe we should mark it as deprecated or something, because it's heinous and since coming up with it, I've realized how bad it is. 

My mantra these days is that you should lead with the data model and work backwards to figure out the code you want. This model of lists that merge together is very much good code leading to an insane (pun intended?) data model, which is just the wrong way to design things.

mlissner commented:
@EvandoBlanco if you haven't used Github Pull Requests before, something to know is that you need to tell your reviewer when things are ready for another look, because they don't usually have a way of telling if a PR is in progress with more work to come or if it's ready for their review. 

I saw you did a force push. Does this mean somebody should take another look at it?

EvandoBlanco commented:
> 
> 
> @EvandoBlanco if you haven't used Github Pull Requests before, something to know is that you need to tell your reviewer when things are ready for another look, because they don't usually have a way of telling if a PR is in progress with more work to come or if it's ready for their review.
> 
> I saw you did a force push. Does this mean somebody should take another look at it?

@mlissner sorry about that. I force pushed because I realized I had pushed without cleaning up my commits and wanted to keep the commit history clean. I think I've resolved all the previously mentioned issues, they only remaining question was with the case_name_shorts. I noticed that whatever auto-generates them doesn't work with the format here so I need to implement that. However, I don't have a lot of domain knowledge about the law (I was drawn to this to learn more about it btw). So I'm not sure how the result would differentiate from what we would expect from _get_case_names(), i.e. _Pennsylvania v. Pepe Silvia , etc._

flooie commented:
@EvandoBlanco Well you've come to the right place to learn more about the law.  We have a handful of case names - short, medium and long depending.  But we try to make those determinations elsewhere.  

Pennsylvania v. Pepe Silvia - is like the classic case_name version.  The full name here would be something like Commonwealth of Pennsylvania v. Pepe Silivia or something like that.  

Case Name Short would probably shorten to Penn v. Silvia.  You it looks like most of the case names on the coast guard website is just case_name

EvandoBlanco commented:
> 
> 
> @EvandoBlanco Well you've come to the right place to learn more about the law. We have a handful of case names - short, medium and long depending. But we try to make those determinations elsewhere.
> 
> Pennsylvania v. Pepe Silvia - is like the classic case_name version. The full name here would be something like Commonwealth of Pennsylvania v. Pepe Silivia or something like that.
> 
> Case Name Short would probably shorten to Penn v. Silvia. You it looks like most of the case names on the coast guard website is just case_name

Good to know. If that's the case I would presume this is good to go.

mlissner commented:
Great. That's to you then, @flooie

flooie commented:
Thanks @mlissner, @EvandoBlanco I'm diving in today.

flooie commented on juriscraper/opinions/united_states/federal_special/uscgcca.py:
```diff
@@ -0,0 +1,76 @@
+        statuses = []
+        for name in self._get_case_names():
+            m = search(publish_regex, name)
+            statuses.append(m.group(0).strip("()- ") if m else "PUBLISHED")
```
PUBLISHED should be Published.

flooie replied:
Actually, this requires a deeper dive.  And it's one that I didn't recognize (mostly) until now.  

<img width="432" alt="Screen Shot 2021-04-13 at 2 36 30 PM" src="https://user-images.githubusercontent.com/6464529/114604376-cde70a00-9c66-11eb-85b0-095d9fa9f589.png">

I'm attaching a screenshot of the documents currently available for help in dissecting this though.

When we say something is published, we don't mean is it available or published online.  We ask is it precedential (having the authority to bind lower courts).  Not all opinions carry the same weight or are binding.  

The easiest way to tell if someone is precedential here is if it gets "reported" in an official reporter.  In this case, I believe the official reporter is the Military Journal (MJ).  You can have a west, lexis, neutral and MJ citation but the MJ citation is basically what denotes that it is Published and precedential.  In other scrapers, you could assign pretty much all full opinions as published if its a Supreme Court for example.  But even there we wouldn't classify orders and minor documents precedential.  

So for the Coast Guard, I think you can check if the document has an MJ citation and also that it is not an opinion.  If those two parts are satisfied, set it as precedential (`Published`).  If it has an MJ citation but is an order, I would hedge and mark it as `Unknown`, and otherwise I would say non-precedential (`Unpublished`)

flooie commented on juriscraper/opinions/united_states/federal_special/uscgcca.py:
```diff
@@ -0,0 +1,76 @@
+            text = name.text_content()
+            m = search(names_regex, text)
+            names.append(m.group(0) if m else text)
+
```
We should remove this extra whitespace

flooie commented on juriscraper/opinions/united_states/federal_special/uscgcca.py:
```diff
@@ -0,0 +1,76 @@
+    def __init__(self, *args, **kwargs):
+        super(Site, self).__init__(*args, **kwargs)
+        self.court_id = self.__module__
+        self.mj_cite_regex = "\\d+ *M\\.*J\\.* *\\d+"
```
This regex doesn't seem to work.  And if you look at the second page of the website you see citations with and without periods.  I think you want something closer to 
`r"\d{1,4} M\.?J\.? \d{1,4}"`

flooie replied:
also I think you should denote all regex strings with an r (r"") in python 3.8+

flooie commented on juriscraper/opinions/united_states/federal_special/uscgcca.py:
```diff
@@ -0,0 +1,76 @@
+        return names
+
+    def _get_precedential_statuses(self):
+        publish_regex = "((- [a-zA-Z]+)|(\\([a-zA-Z ]+\\)))$"
```
Does this regex work?  I dont fully understand it.

flooie commented on juriscraper/opinions/united_states/federal_special/uscgcca.py:
```diff
@@ -0,0 +1,76 @@
+
+    def _get_case_names(self):
+        names_xpath = "//tr//td[1]"
+        names_regex = "^([a-zA-Z ]+ [Vv]\\.? [ ]*[a-zA-Z]+)"
```
This example from the HTML 

`ORDER - In re Tucker Petition 2019 WL 5280525`

Should extract `In re Tucker Petition`

This is a hard thing about regex and legal stuff.

flooie commented on juriscraper/opinions/united_states/federal_special/uscgcca.py:
```diff
@@ -0,0 +1,76 @@
+        super(Site, self).__init__(*args, **kwargs)
+        self.court_id = self.__module__
+        self.mj_cite_regex = "\\d+ *M\\.*J\\.* *\\d+"
+        self.wl_cite_regex = "\\d+ *W\\.*L\\.* *\\d+"
```
This regex pattern also fails.  You add `\\` a couple times when I think you just want `\`.  Check your patterns.  I use https://regex101.com all the time

flooie commented on juriscraper/opinions/united_states/federal_special/uscgcca.py:
```diff
@@ -0,0 +1,76 @@
+        url_xpath = "//tr//td[1]/a/@href"
+        return [requote_uri(url) for url in self.html.xpath(url_xpath)]
+
+    def _get_west_citations(self):
+        return self._parse_citations(self.wl_cite_regex)
+
+    def _get_neutral_citations(self):
+        return self._parse_citations(self.mj_cite_regex)
+
+    def _parse_citations(self, regex):
+        cite_list = []
+        for name in self._get_case_names():
+            m = search(regex, name)
+            cite_list.append(m.group(0) if m else None)
+
+        return cite_list
```
Again, this is from above, I think this could be a nice, but it doesn't work.  I think the problem is at the regex level, but Im not positive of that.

flooie commented on tests/examples/opinions/united_states/uscgcca_example.compare.json:
```diff
@@ -0,0 +1,167 @@
+    "case_name_shorts": ""
+  },
+  {
+    "case_dates": "2019-05-15",
+    "case_names": "UNITED STATES v. CORREA",
+    "download_urls": "https://media.defense.gov/2019/May/21/2002134458/-1/-1/0/UNITED%20STATES%20V.%20CORREA%20%20(MERITS).PDF",
+    "precedential_statuses": "PUBLISHED",
+    "blocked_statuses": false,
+    "date_filed_is_approximate": false,
+    "neutral_citations": null,
+    "west_citations": null,
+    "case_name_shorts": "CORREA"
+  }
```
It's a good practice to check the example compare json to make sure its generating the results you see in the HTML.

flooie reviewed (changes requested):
Hey @EvandoBlanco 

This is another great effort, but I think it still has a bit to go.  Thanks for taking another crack.  I think you probably would've caught most of these things if you checked against the compare.json file.  

It's not obvious and maybe it should be, but when you run the tests for the first time it generates a JSON file to compare your HTML against.  If the parser doesn't parse correctly it generates a bad JSON file to test against.  

I think that's what happened here.  You ran the tests and it generated bad data, but it wasn't failing.  Check the JSON file against the HTML to make sure it's working.  The tests really are to make sure the overall juriscraper structure as much (and maybe more) than individual courts.

I would also check back at some older opinions on the website.  I think a variety of case names etc might inform the final regex patterns.

brianwc commented:
Just wanted to drop a note to say how excited I get when new courts get a scraper and to add that the citation convention for this court is "C.G. Ct. Crim. App." and the M.J. has been the official reporter for them since 1975. So, for example, a full citation would be: _United States v. Weber_, 56 M.J. 736 (C.G. Ct. Crim. App. 2002). 

The other comparable courts are: A. Ct. Crim. App., A.F. Ct. Crim. App., C.G. Ct. Crim. App., N-M. Ct. Crim. App.
(Yes, the Navy and Marines share a court).

I also know that they refer to themselves as CGCCA, which suggests that our slug should be cgcca rather than uscgcca. Also, if we look to those other courts, we find:
A. Ct. Crim. App. = ACCA
A.F. Ct. Crim. App. = AFCCA 
N-M. Ct. Crim. App. = NMCCA

which has no collisions, are the slugs we already adopted for those, and these are the acronyms that these bodies use for themselves, so that's nice. I think all that convinces me that this should just be CGCCA.

flooie commented:
Thanks Brian. 



> On Apr 13, 2021, at 7:26 PM, Brian Carver ***@***.***> wrote:
> 
> ﻿
> Just wanted to drop a note to say how excited I get when new courts get a scraper and to add that the citation convention for this court is "C.G. Ct. Crim. App." and the M.J. has been the official reporter for them since 1975. So, for example, a full citation would be: United States v. Weber, 56 M.J. 736 (C.G. Ct. Crim. App. 2002).
> 
> The other comparable courts are: A. Ct. Crim. App., A.F. Ct. Crim. App., C.G. Ct. Crim. App., N-M. Ct. Crim. App.
> (Yes, the Navy and Marines share a court).
> 
> I also know that they refer to themselves as CGCCA, which suggests that our slug should be cgcca rather than uscgcca. Also, if we look to those other courts, we find:
> A. Ct. Crim. App. = ACCA
> A.F. Ct. Crim. App. = AFCCA
> N-M. Ct. Crim. App. = NMCCA
> 
> which has no collisions, are the slugs we already adopted for those, and these are the acronyms that these bodies use for themselves, so that's nice. I think all that convinces me that this should just be CGCCA.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

EvandoBlanco commented:
@flooie @arderyp @brianwc 
I've taken all the feedback and incorporated it. The suggested precedential status logic is now incorporated. Also, the case name regex has been completely revamped (thanks @flooie for suggesting raw strings, much easier without having to escape everything). Please let me know if any other changes are necessary.

flooie commented:
Thanks @EvandoBlanco.  Cant wait to take a look soon.

flooie commented on juriscraper/opinions/united_states/federal_special/cgcca.py:
```diff
@@ -0,0 +1,108 @@
+        unknown_regex = r".+order.+"
+        statuses = []
+        # If a case has a MJ citation, it is published.
+        # Cases with a MJ citation but are noted as opinions are unpublished.
+        # Cases with a MJ citation but are noted as orders are unknown.
+        for title in self.html.xpath(title_xpath):
+            m = search(self.mj_cite_regex, title.text_content())
+            text = title.text_content()
+            if m:
+                if search(opinion_regex, text, IGNORECASE):
+                    statuses.append("Unpublished")
+                    continue
+                if search(unknown_regex, text, IGNORECASE):
+                    statuses.append("Unknown")
+                    continue
```
This is great.  I caused a bug if your code with a typo in my feedback.  MJ citations + Opinions are published.

EvandoBlanco replied:
So, to be clear, line 79-81:
```
                if search(opinion_regex, text, IGNORECASE):
                    statuses.append("Unpublished")
                    continue
```

just needs to be removed?

EvandoBlanco replied:
@flooie , I've been pretty busy this month but now have time to circle back on this. Is this clarification correct?

flooie commented:
@EvandoBlanco I did a quick once over and I'm excited to do the review later today.  Thanks for taking all the feedback and getting this (not easy) scraper to what looks like a good place.

flooie commented:
@EvandoBlanco  this was pretty dormant (partially my fault) but we are switching to a new streamlined process in juriscraper opinionsitelinear.  I took what you started and implemented it in a new PR.  Hopefully we'll have the coast guard up and running soon.
